repose-aggregator/core/repose-core/src/main/scala/org/openrepose/powerfilter/ReposeFilter.scala

```diff
@@ -19,7 +19,193 @@
+        val tracingHeaderConfig = reposeFilterLoader.getTracingHeaderConfig
+        if (tracingHeaderConfig.exists(_.isRewriteHeader)) wrappedRequest.removeHeader(TRACE_GUID)
+
+        //Grab the traceGUID from the request if there is one, else create one
+        val traceGUID = Option(wrappedRequest.getHeader(TRACE_GUID)) match {
+          case Some(string) if StringUtils.isNotBlank(string) => string
+          case _ => UUID.randomUUID.toString
+        }
+
+        MDC.put(TracingKey.TRACING_KEY, traceGUID)
+
+        try {
+          TryWith(filterContextList)(filterContextListToo => {
+            // Ensure the request URI is a valid URI
+            // This object is only being created to ensure its validity.
+            // So it is safe to suppress warning squid:S1848
+            new URI(wrappedRequest.getRequestURI)
+
+            if (tracingHeaderConfig.exists(_.isEnabled)) {
+              if (StringUtils.isBlank(wrappedRequest.getHeader(TRACE_GUID))) {
+                wrappedRequest.addHeader(TRACE_GUID, TracingHeaderHelper.createTracingHeader(traceGUID, wrappedRequest.getHeader(VIA)))
+              }
+              if (tracingHeaderConfig.exists(_.isSecondaryPlainText)) {
+                TraceIdLogger.trace("Adding plain text trans id to request: {}", traceGUID)
+                wrappedRequest.replaceHeader(REQUEST_ID, traceGUID)
+              }
+              val tracingHeader = wrappedRequest.getHeader(TRACE_GUID)
+              TraceIdLogger.info("Tracing header: {}", TracingHeaderHelper.decode(tracingHeader))
+              wrappedResponse.addHeader(TRACE_GUID, tracingHeader)
+            }
+            new ReposeFilterChain(
+              filterContextListToo.filterContexts,
+              chain,
+              None,
+              optMetricRegistry,
+              tracer
+            ).doFilter(wrappedRequest, wrappedResponse)
+          }) match {
+            case Success(_) =>
+              ThisLogger.trace("{} -- Successfully processed request", nodeId)
+            case Failure(use: URISyntaxException) =>
+              ThisLogger.debug(s"$nodeId -- Invalid URI requested: ${wrappedRequest.getRequestURI}", use)
+              wrappedResponse.sendError(SC_BAD_REQUEST, "Error processing request")
+            case Failure(e: Exception) =>
+              ThisLogger.error(s"$nodeId -- Issue encountered while processing filter chain.", e)
+              wrappedResponse.sendError(SC_BAD_GATEWAY, "Error processing request")
+            case Failure(t: Throwable) =>
+              ThisLogger.error(s"$nodeId -- Error encountered while processing filter chain.", t)
+              wrappedResponse.sendError(SC_BAD_GATEWAY, "Error processing request")
+              throw t
+          }
+        } finally {
+          closeSpan(wrappedResponse, scope)
+          if (!wrappedResponse.isCommitted) {
+            responseHeaderService.setVia(wrappedRequest, wrappedResponse)
```

Comment:
We set the guid before processing (suspect to do), might as well do this as well, and then we don't have to worry about the response being mutable at all.

Reply:
DO we want to uncommit, and then always write the via header?

Reply:
I have been thinking about this one a lot.
As you mentioned it doesn't seem quite right to set it early with the GUID, but that is already being done so we could set this also.
We could also uncommit if able to and set it, but if we can't then it is in the same boat.

I'm torn on which one is right for this case.
@dmnjohns do you have any thoughts on this?

Reply:
Since there is nothing that can be done during processing that would change the outcome of the `setVia` method, I opted to go ahead and move it to before the `ReposeFilterChain.doFilter` method was called.

Reply:
Oops, didn't see this thread a few days ago.

Anyway, there should not be a case where we cannot uncommit if we have wrapped the response in a mutable mode. That would imply that the response was committed before being passed to our filter which should never happen. 

As for nothing being done during processing that would change the outcome of `setVia`... I guess I'm not sure what you mean by that. Clearly setting the `Via` header upfront means that any downstream processing could change/affect it.

While I would love to be able to set the `Via` header upfront (and in doing so drop the need for a wrapped response), doing so is unsafe. For one, we cannot guarantee that the `Via` header makes it out on the response. For two, ordering matters in the `Via` header. If we set it here and forward the request to another proxy which response with a `Via` header, we are going to appear before the downstream proxy in the `Via` header sent back to the user, which violates spec:
https://tools.ietf.org/html/rfc7230#section-5.7.1